### PR TITLE
Soft delete vaults

### PR DESCRIPTION
**Pull request by flvndvd**
## Description

This PR fixes https://github.com/dust-tt/tasks/issues/1443.

This PR introduces soft deletion for vaults, which necessitates removing all associated resources:
- Data source views
- Data sources
- Apps
- Keys
- Groups

I've streamlined the various deletion processes, consolidating them into a unified approach. The deletion logic is now located in `lib/api/<resource>.ts` and includes both a `softDeleteXXXXAndLaunchWorkflow` method and a `hardDeleteXXX` method.

## Risk

Worst case it breaks deletions. Safe to rollback.

## Deploy Plan

- [x] Apply SQL migration
- [ ] Deploy Front
- [ ] Restart NDE scrub workflows

### CODE REVIEW

File: front/lib/api/data_sources.ts [hunk 1]
```diff
@@ -89,12 +90,7 @@ export async function hardDeleteDataSource(
   auth: Authenticator,
   dataSource: DataSourceResource
 ) {
-  if (!auth.isBuilder()) {
-    return new Err({
-      code: "unauthorized_deletion",
-      message: "Only builders can destroy data sources.",
-    });
-  }
+  assert(auth.isBuilder(), "Only builders can delete data sources.");
```
Comment (tdraier) [inline front/lib/api/data_sources.ts:93]: are we ok to replace a Result by a throw here ? why not in `softDeleteDataSourceAndLaunchScrubWorkflow` then ?

File: front/lib/api/vaults.ts [hunk 2]
```diff
@@ -59,13 +58,13 @@ export const deleteVault = async (
   if (groupHasKeys > 0) {
     return new Err(
       new Error(
-        "Canno't delete group with active API Keys. Please revoke all keys before."
+        "Can not delete group with active API Keys. Please revoke all keys before."
```
Comment (tdraier) [inline front/lib/api/vaults.ts:61]: nit: "Cannot"  🙃

File: front/lib/resources/vault_resource.ts [hunk 3]
```diff
@@ -236,68 +247,34 @@ export class VaultResource extends BaseResource<VaultModel> {
...
       where: {
-        id: {
-          [Op.in]: vaultIds,
-        },
+        id: this.id,
       },
       transaction,
+      // Soft delete the vault.
+      hardDelete: false,
     });
```
Comment (tdraier) [inline front/lib/resources/vault_resource.ts:275]: Can't we just have 
```
 await VaultModel.destroy({
      where: {
        id: this.id,
      },
      transaction,
      hardDelete,
    });
```
?

File: front/lib/api/data_sources.ts (hunk 1 above)
Comment (flvndvd) [reply front/lib/api/data_sources.ts:93]: 👀

Comment (flvndvd) [reply front/lib/api/data_sources.ts:93]: So `softDeleteXXXX` can be called from the API(s) hence the need to have proper handling, `hardDeleteXXX` on the other side, should only be called from scrubbing workflows.

File: front/lib/resources/vault_resource.ts (hunk 3 above)
Comment (flvndvd) [reply front/lib/resources/vault_resource.ts]: 👍

**Review by tdraier — APPROVED** [review]

File: front/lib/api/vaults.ts [hunk 4]
```diff
@@ -77,27 +76,75 @@ export const deleteVault = async (
       }
     }
 
+    // Soft delete data sources they will be hard deleted in the scrubbing job.
     for (const ds of dataSources) {
-      const res = await softDeleteDataSourceAndLaunchScrubWorkflow(auth, ds, t);
+      const res = await ds.delete(auth, { hardDelete: false, transaction: t });
```
Comment (Fraggle) [inline front/lib/api/vaults.ts:81]: so should we use a dedicated method or delete(.., {hardDelete: false}) ?

(i rather have one way)

File: front/lib/api/vaults.ts [hunk 5]
```diff
@@ -1,25 +1,24 @@
...
-): Promise<Result<undefined, Error>> => {
-  if (!auth.isAdmin()) {
-    throw new Error("Only admins can delete vaults.");
-  }
-  if (!vault.isRegular()) {
-    throw new Error("Cannot delete non regular vaults.");
-  }
+) {
+  assert(auth.isAdmin(), "Only admins can delete vaults.");
+  assert(vault.isRegular(), "Cannot delete non regular vaults.");
```
[the remark this answers is not in the document]
Comment (flvndvd) [reply front/lib/api/vaults.ts:21]: For preconditions or when code is here mainly to please TS, assert is shorter. In regular TS throwing means that the code sometimes know how to handle those exception, here asserting that this, is definitely meant to NEVER happen.

**Review by Fraggle — APPROVED** [review]
LGTM.

**Pull request merged**